### PR TITLE
Fix Multi-Agent Episode concatenation for sequential environments

### PR DESCRIPTION
**Pull request by pseudo-rnd-thoughts**
## Description
For https://github.com/ray-project/ray/pull/59508 and https://github.com/ray-project/ray/pull/59581 that using TicTacToe would cause the following error

```
ray::DQN.train() (pid=88183, ip=127.0.0.1, actor_id=2b775f13e808cc4aaaa23bde01000000, repr=DQN(env=<class 'ray.rllib.examples.envs.classes.multi_agent.tic_tac_toe.TicTacToe'>; env-runners=0; learners=0; multi-agent=True))
  File "ray/python/ray/tune/trainable/trainable.py", line 331, in train
    raise skipped from exception_cause(skipped)
  File "ray/python/ray/tune/trainable/trainable.py", line 328, in train
    result = self.step()
  File "ray/python/ray/rllib/algorithms/algorithm.py", line 1242, in step
    train_results, train_iter_ctx = self._run_one_training_iteration()
  File "ray/python/ray/rllib/algorithms/algorithm.py", line 3666, in _run_one_training_iteration
    training_step_return_value = self.training_step()
  File "ray/python/ray/rllib/algorithms/dqn/dqn.py", line 646, in training_step
    return self._training_step_new_api_stack()
  File "ray/python/ray/rllib/algorithms/dqn/dqn.py", line 668, in _training_step_new_api_stack
    self.local_replay_buffer.add(episodes)
  File "ray/python/ray/rllib/utils/replay_buffers/prioritized_episode_buffer.py", line 314, in add
    existing_eps.concat_episode(eps)
  File "ray/python/ray/rllib/env/multi_agent_episode.py", line 862, in concat_episode
    sa_episode.concat_episode(other.agent_episodes[agent_id])
  File "ray/python/ray/rllib/env/single_agent_episode.py", line 618, in concat_episode
    assert self.t == other.t_started
AssertionError
```

In the multi-agent-episode `concat_episode`, we check if any agent hasn't received their next observation from an observation, action. This results in a hanging action where in one episode is the observation, action then in the next is the resulting observation, reward, etc. This [code](https://github.com/ray-project/ray/blob/22cf6ef6af2cddc233bca7ce59668ed8f4bbb17e/rllib/env/multi_agent_episode.py#L848) check if this has happened then added an extra step at the beginning to include this hanging data. 

However, in testing, the multi-agent episode `cut` method already implements this (if using `slice` this will cause a hidden bug) meaning that an extra unnecessary step's data is being added resulting in the episode beginnings not lining up. 
Therefore, this PR removes this code and replaces with a simple check to assume that the hanging action is equivalent to the initial action in the next episode. 

For testing, I found that the `concat_episode` test was using `slice` which doesn't account for hanging data while `cut` which is used in the env-runner does. I modified the test to be more functional based where I created a custom environment that has agents taking actions at different frequencies then returning as an observation the agent's timestep. This allows us to test through concatenating all episodes of the same ID and checking that the observations increase 0, 1, 2, ... and ensures that no data goes missing for users.

### CODE REVIEW

File: rllib/connectors/common/numpy_to_tensor.py [hunk 1]
```diff
@@ -111,7 +111,8 @@ def __call__(
                     )
                 else:
                     raise ValueError(
-                        "`NumpyToTensor`does NOT support frameworks other than torch!"
+                        "`NumpyToTensor`does NOT support frameworks other than torch! "
```
Comment (simonsays1980) [inline rllib/connectors/common/numpy_to_tensor.py:114]: Nice!

File: rllib/env/tests/test_multi_agent_episode.py [hunk 2]
```diff
@@ -125,6 +131,128 @@ def step(
...
+                self.agent_timestep[agent] += 1
+
+        assert obs, f"{obs=}, {self.env_timestep=}, {self.agent_timestep=}"
+        return obs
+
+
+def create_env(config):
+    return SequentialMultiAgentEnv(
+        agent_fns={
+            "p_true": lambda x: True,
```
Comment (simonsays1980) [inline rllib/env/tests/test_multi_agent_episode.py:231]: Nice idea!!

File: rllib/env/multi_agent_episode.py [hunk 3]
```diff
@@ -798,7 +798,9 @@ def concat_episode(self, other: "MultiAgentEpisode") -> None:
         """Adds the given `other` MultiAgentEpisode to the right side of `self`.
 
         In order for this to work, both chunks (`self` and `other`) must fit
-        together. This is checked by the IDs (must be identical), the time step counters
+        together that are split through `cut`. For sequential multi-agent environments
+        using slice might cause problems from hanging observation/actions.
```
Comment (simonsays1980) [inline rllib/env/multi_agent_episode.py:802]: This is something we need to fix in the near future. Could you raise another issue on Ray OSS please?

File: rllib/env/multi_agent_episode.py [hunk 4]
```diff
@@ -798,7 +798,9 @@ def concat_episode(self, other: "MultiAgentEpisode") -> None:
...
+        This is checked by the IDs (must be identical), the time step counters
         (`self.env_t` must be the same as `episode_chunk.env_t_started`), as well as the
```
Comment (simonsays1980) [inline rllib/env/multi_agent_episode.py]: `episode_chunk.env_t_started` -> `other.env_t_started`

**Review by simonsays1980 — CHANGES_REQUESTED** [review]
Thanks for fixing this @pseudo-rnd-thoughts ! Small things: could you raise an issue for the `slice` method to be fixed in the future? And it needs a small change in the docstring -  did not came from your change though

File: rllib/env/multi_agent_episode.py (hunk 3 above)
Comment (pseudo-rnd-thoughts) [reply rllib/env/multi_agent_episode.py:802]: I'm not sure if this is a bug or an inherent limitation of the `slice` method

File: rllib/env/multi_agent_episode.py (hunk 4 above)
Comment (pseudo-rnd-thoughts) [reply rllib/env/multi_agent_episode.py]: Done

File: rllib/env/single_agent_episode.py [hunk 5]
```diff
@@ -598,7 +598,7 @@ def concat_episode(self, other: "SingleAgentEpisode") -> None:
 
         In order for this to work, both chunks (`self` and `other`) must fit
         together. This is checked by the IDs (must be identical), the time step counters
-        (`self.env_t` must be the same as `episode_chunk.env_t_started`), as well as the
+        (`self.env_t` must be the same as `other.env_t_started`), as well as the
```
Comment (simonsays1980) [inline rllib/env/single_agent_episode.py:601]: Thanks!

**Review by simonsays1980 — APPROVED** [review]
LGTM. Thanks @pseudo-rnd-thoughts !
